lib/spack/spack/cmd/external.py

```diff
@@ -190,163 +86,6 @@ def external_find(args):
-    # prefix
-    assert os.path.isdir(prefix)
-    if os.path.basename(prefix) == 'bin':
-        return os.path.dirname(prefix)
-
-
-def _get_predefined_externals():
-    # Pull from all scopes when looking for preexisting external package
-    # entries
-    pkg_config = spack.config.get('packages')
-    already_defined_specs = set()
-    for pkg_name, per_pkg_cfg in pkg_config.items():
-        for item in per_pkg_cfg.get('externals', []):
-            already_defined_specs.add(spack.spec.Spec(item['spec']))
-    return already_defined_specs
-
-
-def _update_pkg_config(scope, pkg_to_entries, not_buildable):
-    predefined_external_specs = _get_predefined_externals()
-
-    pkg_to_cfg, all_new_specs = {}, []
-    for pkg_name, ext_pkg_entries in pkg_to_entries.items():
-        new_entries = list(
-            e for e in ext_pkg_entries
-            if (e.spec not in predefined_external_specs))
-
-        pkg_config = _generate_pkg_config(new_entries)
-        all_new_specs.extend([
-            spack.spec.Spec(x['spec']) for x in pkg_config.get('externals', [])
-        ])
-        if not_buildable:
-            pkg_config['buildable'] = False
-        pkg_to_cfg[pkg_name] = pkg_config
-
-    pkgs_cfg = spack.config.get('packages', scope=scope)
-
-    pkgs_cfg = spack.config.merge_yaml(pkgs_cfg, pkg_to_cfg)
-    spack.config.set('packages', pkgs_cfg, scope=scope)
-
-    return all_new_specs
-
-
-def _get_external_packages(packages_to_check, system_path_to_exe=None):
-    if not system_path_to_exe:
-        system_path_to_exe = _get_system_executables()
-
-    exe_pattern_to_pkgs = defaultdict(list)
-    for pkg in packages_to_check:
-        if hasattr(pkg, 'executables'):
-            for exe in pkg.executables:
-                exe_pattern_to_pkgs[exe].append(pkg)
-
-    pkg_to_found_exes = defaultdict(set)
-    for exe_pattern, pkgs in exe_pattern_to_pkgs.items():
-        compiled_re = re.compile(exe_pattern)
-        for path, exe in system_path_to_exe.items():
-            if compiled_re.search(exe):
-                for pkg in pkgs:
-                    pkg_to_found_exes[pkg].add(path)
-
-    pkg_to_entries = defaultdict(list)
-    resolved_specs = {}  # spec -> exe found for the spec
-
-    for pkg, exes in pkg_to_found_exes.items():
-        if not hasattr(pkg, 'determine_spec_details'):
-            tty.warn("{0} must define 'determine_spec_details' in order"
-                     " for Spack to detect externally-provided instances"
-                     " of the package.".format(pkg.name))
-            continue
-
-        # TODO: iterate through this in a predetermined order (e.g. by package
-        # name) to get repeatable results when there are conflicts. Note that
-        # if we take the prefixes returned by _group_by_prefix, then consider
-        # them in the order that they appear in PATH, this should be sufficient
-        # to get repeatable results.
```

Comment:
When I read this TODO I thought that any arbitrary sorting would do to fix this, so I removed the TODOs and added:
```python
for prefix, exes_in_prefix in sorted(_group_by_prefix(exes)):
```
The second TODO seemed misplaced or redundant. We can add extra attributes to detected packages, so any detail to differentiate them can be encoded there I guess.

Reply:
I agree with you that the sorting you added should resolve the first TODO, although Ideally we would order the prefixes based on where they appear in PATH (when possible), because users may expect to be able to influence the search by adjusting their PATH (if two executables report the same spec, then only the first one visited will produce an entry).

> The second TODO seemed misplaced or redundant. We can add extra attributes to detected packages, so any detail to differentiate them can be encoded there I guess.

Until we do that, I think it's worth recording that we don't handle it properly. As it is, if you have executables like `gcc-7` and `gcc-8` in the same directory, then for now we don't provide sufficient information to distinguish them (both would have the base directory as the prefix). IMO that was clearest here, but if you think there's a better location to mention that then that also works with me.

Reply:
Restored the second TODO in 8f789d4